java/src/main/java/com/xebia/functional/xef/java/auto/AIScope.java

```diff
@@ -0,0 +1,147 @@
+import com.fasterxml.jackson.module.jsonSchema.JsonSchemaGenerator;
+import com.xebia.functional.loom.LoomAdapter;
+import com.xebia.functional.xef.AIError;
+import com.xebia.functional.xef.auto.CoreAIScope;
+import com.xebia.functional.xef.embeddings.Embeddings;
+import com.xebia.functional.xef.embeddings.OpenAIEmbeddings;
+import com.xebia.functional.xef.env.OpenAIConfig;
+import com.xebia.functional.xef.llm.openai.KtorOpenAIClient;
+import com.xebia.functional.xef.llm.openai.LLMModel;
+import com.xebia.functional.xef.llm.openai.functions.CFunction;
+import com.xebia.functional.xef.llm.openai.images.ImageGenerationUrl;
+import com.xebia.functional.xef.llm.openai.images.ImagesGenerationResponse;
+import com.xebia.functional.xef.textsplitters.TextSplitter;
+import com.xebia.functional.xef.vectorstores.LocalVectorStore;
+import com.xebia.functional.xef.vectorstores.VectorStore;
+import kotlin.collections.CollectionsKt;
+import kotlin.jvm.functions.Function1;
+import com.xebia.functional.xef.pdf.PDFLoaderKt;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+
+public class AIScope implements AutoCloseable {
+    private CoreAIScope coreAIScope;
+    private ObjectMapper om;
+    private JsonSchemaGenerator schemaGen;
+
+    public AIScope(CoreAIScope coreAIScope) {
+        this.coreAIScope = coreAIScope;
+        this.om = new ObjectMapper();
+        this.schemaGen = new JsonSchemaGenerator(om);
+    }
+
+    public AIScope(CoreAIScope coreAIScope, ObjectMapper om) {
+        this.coreAIScope = coreAIScope;
+        this.om = om;
+        this.schemaGen = new JsonSchemaGenerator(om);
+    }
+
+    public static <T> T run(Function1<AIScope, T> block) {
+        OpenAIConfig config = new OpenAIConfig();
+        KtorOpenAIClient client = new KtorOpenAIClient(config);
+        try {
+            Embeddings embeddings = new OpenAIEmbeddings(config, client);
+            VectorStore vectorStore = new LocalVectorStore(embeddings);
+            CoreAIScope scope = new CoreAIScope(LLMModel.getGPT_3_5_TURBO(), LLMModel.getGPT_3_5_TURBO_FUNCTIONS(), client, vectorStore, embeddings, 3, "user", false, 0.4, 1, 20, 500);
+            return block.invoke(new AIScope(scope));
+        } finally {
+            client.close();
+        }
+    }
+
+    private <T> T undefined() {
+        throw new RuntimeException("Method is undefined");
+    }
+
+    public <A> A prompt(String prompt, Class<A> cls) {
+        return prompt(prompt, cls, coreAIScope.getMaxDeserializationAttempts(), coreAIScope.getDefaultSerializationModel(), coreAIScope.getUser(), coreAIScope.getEcho(), coreAIScope.getNumberOfPredictions(), coreAIScope.getTemperature(), coreAIScope.getDocsInContext(), coreAIScope.getMinResponseTokens());
+    }
+
+    public <A> A prompt(String prompt, Class<A> cls, Integer maxAttempts, LLMModel llmModel, String user, Boolean echo, Integer n, Double temperature, Integer bringFromContext, Integer minResponseTokens) {
```

Comment:
I didn't know that `record` was a valid keyword in Java. I can use it in further examples rather than using public classes with public static fields 😄

Reply:
Word of caution, though, is that this feature is only available in versions after Java 16, so perhaps using it would restrict which JVM versions can use this library.

Is there a decision as to what versions of the JVM would this be targetting?

Reply:
@diesalbla, good point. We were planning to remove the Loom requirement as well, but perhaps we can just choose a higher JDK version instead. JDK 21 is around the corner, but perhaps a bit too ambitious. Having LOOM would be quite neat though, could get some _async_ hurdles out of the way.

Reply:
See discussion on Slack, we decided on JDK8. https://fortyseven.slack.com/archives/C051PG86AH0/p1687795788974039?thread_ts=1687789966.900839&cid=C051PG86AH0